impl: Store Manager Order History Page

London-Alex opened the pull request:
This branch implements a View Order History page for store managers. The page will display a message if the user is not logged in as a store manager or if the store has no orders to display. If a store has orders it shows a table with order summaries and then more detailed order info below the table.

The following files were changed:
- shop.route: Provides the API route to OrderController.getStoreOrders().
- order.controller: Calls OrderDAO.findStoreOrders() to get orders and then processes the orders to remove other stores' items and recalculate prices.
- orderDAO: Queries database to find all orders with items belonging to that store manager.
- userAction: Create the frontend function to call the API.
- App: Add the route for the StoreOrdersScreen.
- header: Add a dropdown option in the store manager menu to link to the screen.
- StoreOrdersScreen: Call the API to get the orders, then process and display the order info.

GitHub Issue #73

London-Alex committed:
Merge remote-tracking branch 'origin/ke-sandra-implement-place-order-function' into alex-store-manager-order-history

London-Alex committed:
impl: Retrieve a Customer's Order History From Backend

This commit implements the backend functions and route necessary to query the database and retrieve a customer's order history.

It also add the frontend user action to send the GET request to the backend.

GitHub Issue #72

London-Alex committed:
impl: Retrieve a Store's Order History From Backend

This commit implements the backend functions and route necessary to query the database and retrieve a store's order history and return only the items from orders that belong to that store.

It also add the frontend user action to send the GET request to the backend.

GitHub Issue #73

London-Alex committed:
impl: Store Manager Order History Page

A previous commit implemented the backend API for fetching orders and returning the info to the frontend. This commit implements the frontend necessary to process and display the store's order history info page.

-App.js was updated to add the route to the page.

-header.js was updated to add a dropdown option to the Store Manager dropdown menu that links to the page.

- StoreOrdersScreen was created to display the store's order history.

The page will display a message if the user is not logged in as a store manager or if the store has no orders to display. If a store has orders it shows a table with order summaries and then more detailed order info below the table.

GitHub Issue #73

London-Alex committed:
refactor: Combine Code to Process Store's Order

A store requires an order from the database to be reprocessed in two regards before being displayed on the store manager's order history page.

1. If an order has items from multiple stores, the items in the order that do not belong to the store should be removed from the order so that a store manager only sees the items ordered from their store.
2. The total price of the order has to be recalculated to include only the price of that store's items, rather than the price of the whole order.

The first aspect was originally handled in the backend before returning to the frontend, and the second aspect was originally handled in the frontend. This created some duplicate code because both required nested for loops to go through each order and each item in each order.

I reduced some of the repeat code by combining both parts into the backend.

GitHub Issue #73

London-Alex committed:
Merge branch 'main' into alex-store-order-history

London-Alex committed:
impl: Add Payment Info to Orders

This commit adds an additional column to the table that contains the Stripe payment_method_id to show the store manager that the payment has been processed and to provide the id of the payment so they could check it.

GitHub Issue #73

SuperRedTrain reviewed (changes requested):
The PR appropriately labelled and has detailed description in the first comment. It also has the appropriate issue linked. I tested this branch on my local machine and a manager order history displayed as described. There is only one small issue: It shows a warning on the console: "Warning: Each child in a list should have a unique "key" prop." I think a "key" should be added in the StoreOrdersScreen.js. Then I will approve the PR.

London-Alex committed:
Fix: Remove Unique Key Prop Warning

The console was displaying a "Warning: Each child in a list should have a unique "key" prop." warning. Resolving this required adding key={order._id} or key={item._id} within the map functions. This assigned unique keys to each child of each list.

The shorthand <> had to be updated to the explicit <React.Fragment> in order to assign a key value.

GitHub Issue #73

SuperRedTrain reviewed (approved):
The issue was solved. Thank you for your work, Alex.

London-Alex committed:
Merge branch 'main' into alex-store-order-history
